internal/services/cdn/cdn_frontdoor_helpers.go

```diff
@@ -439,6 +484,49 @@ func updateRouteAssociations(d *pluginsdk.ResourceData, meta interface{}, routeI
 	return nil
 }
 
+func setRouteRuleSetAssociations(d *pluginsdk.ResourceData, meta interface{}, id *parse.FrontDoorRuleSetAssociationId, ruleSets []interface{}, props *cdn.RouteProperties, errorTxt string, futureErrorTxt string) error {
+	client := meta.(*clients.Client).Cdn.FrontDoorRoutesClient
+	workaroundsClient := azuresdkhacks.NewCdnFrontDoorRoutesWorkaroundClient(client)
+	ctx, routeCancel := timeouts.ForUpdate(meta.(*clients.Client).StopContext, d)
+	defer routeCancel()
+
+	// NOTE: You need to always pass these these three on update else you will
+	// disable/disassociate your custom domains, cache configuration or rule sets...
+	updateProps := azuresdkhacks.RouteUpdatePropertiesParameters{
+		CustomDomains:      props.CustomDomains,
+		CacheConfiguration: props.CacheConfiguration,
+	}
+
+	// Since delete is the only function that passes nil as the ruleSet value
+	// we know it is a delete operation...
+	if ruleSets != nil || len(ruleSets) != 0 {
+		// associate only the rule sets contained within the association resource
+		updateProps.RuleSets = expandRuleSetReferenceArray(ruleSets)
+	} else {
+		// Disassociate all rule sets from the route
+		updateProps.RuleSets = nil
+	}
+
+	updateParams := azuresdkhacks.RouteUpdateParameters{
+		RouteUpdatePropertiesParameters: &updateProps,
+	}
+
+	// lock the route resource for update...
+	locks.ByName(id.AssociationName, cdnFrontDoorRouteResourceName)
+	defer locks.UnlockByName(id.AssociationName, cdnFrontDoorRouteResourceName)
```

Comment:
as discussed in the sync, we likely want to move this up to the top of the Create/Update/Delete methods and lock on both the Association name but also the FrontDoor (or FrontDoor Route) name - to ensure that only one change is happening in parallel?

Reply:
Fixed.